From 09bf8668325ab17bbefe16efcdbf286a4c4c4770 Mon Sep 17 00:00:00 2001
From: acidmanifesto <joshua.lee.betts@gmail.com>
Date: Mon, 8 Feb 2021 16:27:24 -0500
Subject: [PATCH] Death Knight Skip Module (#37)

* Death Knight Skip Module

Skip Death Knight Module
Original Module From Single Player Project Consolidated Skip Module
Rewritten for TC 335 By Single Player Project Developer MDic
Original Concept from conanhun513
Works for both horde and alliance new characters.
When Module is active (Conf = 1) and then a new character Death Knight is created, intro is played (not skippable). Character is then given flags and triggers based on quest completions for the starter areas and give the initial turn in quest after leaving the ebon hold. Alliance DK will be in storm wind with the quest https://www.wowhead.com/quest=13188/where-kings-walk ready for turn in, and Horde DK will be in Orgrimmar with the quest https://www.wowhead.com/quest=13189/saurfangs-blessing. All obtainable key Character Items are in the player's inventory or auto equipped, (Tabard, Weapons, Armors, and such). Spells such as Death Gate works and does return you to the appropriate ebon hold after the DK start zone chain quests are finished.
Runeforging, journeyman riding skipp and deathcharge mount is learned as well.
Tested and working on current 335 with all items and quests and triggers researched for 335 branch.

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

* Fix white space

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

* more whitespace removal.

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

* adjust null to nullptr

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

* NULL to false

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

* conf correction and announcement module addon

conf correction, announcement module addon, and internal description correction

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

* shift and correction

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

* Announcement Shift, GM DK Conf and DK start level configurable

Shifted Announcement Module to its own void.
Added Death Knight Configurable Start Level via conf
Added option for module to be active for GM accounts only.

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

* blank line removal

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

* Corrected Description.

Portable Rune Forge was a Legion V2 feature that is not possible on 334.

Co-Authored-By: ConanHUN <conan_513@hotmail.com>
Co-Authored-By: Serithul <62258982+serithul@users.noreply.github.com>
Co-Authored-By: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>

Co-authored-by: ConanHUN <conan_513@hotmail.com>
Co-authored-by: Serithul <62258982+serithul@users.noreply.github.com>
Co-authored-by: Warlord Zul'Jin <42502821+takasi689@users.noreply.github.com>
---
 src/server/scripts/Custom/SkipDK.cpp          | 165 ++++++++++++++++++
 .../scripts/Custom/custom_script_loader.cpp   |   3 +
 src/server/worldserver/worldserver.conf.dist  |  40 +++++
 3 files changed, 208 insertions(+)
 create mode 100644 src/server/scripts/Custom/SkipDK.cpp

diff --git a/src/server/scripts/Custom/SkipDK.cpp b/src/server/scripts/Custom/SkipDK.cpp
new file mode 100644
index 0000000000..4a0c250b5b
--- /dev/null
+++ b/src/server/scripts/Custom/SkipDK.cpp
@@ -0,0 +1,165 @@
+/* Skip Death Knight Module
+   Original Module From Single Player Project Consolidated Skip Module
+   Rewritten for TC 335 By Single Player Project Developer MDic
+   Original Concept from conanhun513
+*/
+
+#include "AccountMgr.h"
+#include "ScriptMgr.h"
+#include "Player.h"
+#include "Config.h"
+#include "Common.h"
+#include "Chat.h"
+#include "ObjectAccessor.h"
+#include "ObjectMgr.h"
+#include "Player.h"
+#include "SharedDefines.h"
+#include "World.h"
+#include "WorldSession.h"
+
+class Trinitycore_skip_deathknight_announce : public PlayerScript
+{
+public:
+    Trinitycore_skip_deathknight_announce() : PlayerScript("Trinitycore_skip_deathknight_announce") { }
+
+    void OnLogin(Player* Player, bool /*firstLogin*/) override
+    {
+        if (sConfigMgr->GetBoolDefault("Skip.Deathknight.Starter.Announce.enable", true))
+        {
+            ChatHandler(Player->GetSession()).SendSysMessage("This server is running the |cff4CFF00Trinitycore Skip Deathknight Starter |rmodule.");
+        }
+    }
+};
+
+class Trinitycore_skip_deathknight : public PlayerScript
+{
+public:
+    Trinitycore_skip_deathknight() : PlayerScript("Trinitycore_skip_deathknight") { }
+
+    void OnLogin(Player* player, bool firstLogin) override
+    {
+        int DKL = sConfigMgr->GetFloatDefault("Skip.Deathknight.Start.Level", 58);
+
+        if (sConfigMgr->GetBoolDefault("Skip.Deathknight.Starter.Enable", true))
+        {
+            if (player->GetAreaId() == 4342)
+            {
+                if (!firstLogin)
+                    return;
+                player->SetLevel(DKL);
+                player->LearnSpell(53428, false);//runeforging
+                player->LearnSpell(53441, false);//runeforging
+                player->LearnSpell(53344, false);//runeforging
+                player->LearnSpell(62158, false);//runeforging
+                player->LearnSpell(33391, false);//journeyman riding
+                player->LearnSpell(54586, false);//runeforging credit
+                player->LearnSpell(48778, false);//acherus deathcharger
+                player->LearnSkillRewardedSpells(776, 375);//Runeforging
+                player->LearnSkillRewardedSpells(960, 375);//Runeforging
+                player->EquipNewItem(EQUIPMENT_SLOT_HEAD, 38661, true);//Greathelm of the Scourge Champion
+                player->EquipNewItem(EQUIPMENT_SLOT_WRISTS, 38666, true);//Plated Saronite Bracers
+                player->EquipNewItem(EQUIPMENT_SLOT_WAIST, 38668, true);//The Plaguebringer's Girdle
+                player->EquipNewItem(EQUIPMENT_SLOT_HANDS, 38667, true);//Bloodbane's Gauntlets of Command
+                player->EquipNewItem(EQUIPMENT_SLOT_CHEST, 38665, true);//Saronite War Plate
+                player->EquipNewItem(EQUIPMENT_SLOT_LEGS, 38669, true);//Engraved Saronite Legplates
+                player->EquipNewItem(EQUIPMENT_SLOT_SHOULDERS, 38663, true);// Blood-Soaked Saronite Plated Spaulders
+                player->EquipNewItem(EQUIPMENT_SLOT_FEET, 38670, true);//Greaves of the Slaughter
+                player->EquipNewItem(EQUIPMENT_SLOT_TRINKET1, 38675, true);//Signet of the Dark Brotherhood
+                player->EquipNewItem(EQUIPMENT_SLOT_TRINKET2, 38674, true);//Soul Harvester's Charm
+                player->EquipNewItem(EQUIPMENT_SLOT_FINGER1, 38671, true);//Valanar's Signet Ring
+                player->EquipNewItem(EQUIPMENT_SLOT_FINGER2, 38672, true);// Keleseth's Signet Ring
+                player->AddItem(39320, true);//Sky Darkener's Shroud of Blood
+                player->AddItem(38664, true);//Sky Darkener's Shroud of the Unholy
+                player->AddItem(39322, true);//Shroud of the North Wind
+                player->AddItem(38632, true);//Greatsword of the Ebon Blade
+                player->AddItem(6948, true);//Hearthstone
+                player->AddItem(38707, true);//Runed Soulblade
+                player->AddItem(40483, true);//Insignia of the Scourge
+
+                if (player->GetQuestStatus(12657) == QUEST_STATUS_NONE)//The Might Of The Scourge
+                {
+                    player->AddQuest(sObjectMgr->GetQuestTemplate(12657), nullptr);
+                    player->RewardQuest(sObjectMgr->GetQuestTemplate(12657), false, player);
+                }
+                if (player->GetQuestStatus(12801) == QUEST_STATUS_NONE)//The Light of Dawn
+                {
+                    player->AddQuest(sObjectMgr->GetQuestTemplate(12801), nullptr);
+                    player->RewardQuest(sObjectMgr->GetQuestTemplate(12801), false, player);
+                }
+                if (player->GetTeam() == ALLIANCE && player->GetQuestStatus(13188) == QUEST_STATUS_NONE)//Where Kings Walk
+                    player->AddQuest(sObjectMgr->GetQuestTemplate(13188), nullptr);
+                else if (player->GetTeam() == HORDE && player->GetQuestStatus(13189) == QUEST_STATUS_NONE)//Saurfang's Blessing
+                    player->AddQuest(sObjectMgr->GetQuestTemplate(13189), nullptr);
+                if (player->GetTeam() == ALLIANCE)
+                    player->TeleportTo(0, -8833.37f, 628.62f, 94.00f, 1.06f);//Stormwind
+                else
+                    player->TeleportTo(1, 1569.59f, -4397.63f, 16.06f, 0.54f);//Orgrimmar
+                ObjectAccessor::SaveAllPlayers();//Save
+            }
+        }
+
+        if (sConfigMgr->GetBoolDefault("GM.Skip.Deathknight.Starter.Enable", true))
+        {
+            if (player->GetSession()->GetSecurity() >= SEC_MODERATOR && player->GetAreaId() == 4342)
+            {
+                if (!firstLogin)
+                    return;
+                player->SetLevel(DKL);
+                player->LearnSpell(53428, false);//runeforging
+                player->LearnSpell(53441, false);//runeforging
+                player->LearnSpell(53344, false);//runeforging
+                player->LearnSpell(62158, false);//runeforging
+                player->LearnSpell(33391, false);//journeyman riding
+                player->LearnSpell(54586, false);//runeforging credit
+                player->LearnSpell(48778, false);//acherus deathcharger
+                player->LearnSkillRewardedSpells(776, 375);//Runeforging
+                player->LearnSkillRewardedSpells(960, 375);//Runeforging
+                player->EquipNewItem(EQUIPMENT_SLOT_HEAD, 38661, true);//Greathelm of the Scourge Champion
+                player->EquipNewItem(EQUIPMENT_SLOT_WRISTS, 38666, true);//Plated Saronite Bracers
+                player->EquipNewItem(EQUIPMENT_SLOT_WAIST, 38668, true);//The Plaguebringer's Girdle
+                player->EquipNewItem(EQUIPMENT_SLOT_HANDS, 38667, true);//Bloodbane's Gauntlets of Command
+                player->EquipNewItem(EQUIPMENT_SLOT_CHEST, 38665, true);//Saronite War Plate
+                player->EquipNewItem(EQUIPMENT_SLOT_LEGS, 38669, true);//Engraved Saronite Legplates
+                player->EquipNewItem(EQUIPMENT_SLOT_SHOULDERS, 38663, true);// Blood-Soaked Saronite Plated Spaulders
+                player->EquipNewItem(EQUIPMENT_SLOT_FEET, 38670, true);//Greaves of the Slaughter
+                player->EquipNewItem(EQUIPMENT_SLOT_TRINKET1, 38675, true);//Signet of the Dark Brotherhood
+                player->EquipNewItem(EQUIPMENT_SLOT_TRINKET2, 38674, true);//Soul Harvester's Charm
+                player->EquipNewItem(EQUIPMENT_SLOT_FINGER1, 38671, true);//Valanar's Signet Ring
+                player->EquipNewItem(EQUIPMENT_SLOT_FINGER2, 38672, true);// Keleseth's Signet Ring
+                player->AddItem(39320, true);//Sky Darkener's Shroud of Blood
+                player->AddItem(38664, true);//Sky Darkener's Shroud of the Unholy
+                player->AddItem(39322, true);//Shroud of the North Wind
+                player->AddItem(38632, true);//Greatsword of the Ebon Blade
+                player->AddItem(6948, true);//Hearthstone
+                player->AddItem(38707, true);//Runed Soulblade
+                player->AddItem(40483, true);//Insignia of the Scourge
+
+                if (player->GetQuestStatus(12657) == QUEST_STATUS_NONE)//The Might Of The Scourge
+                {
+                    player->AddQuest(sObjectMgr->GetQuestTemplate(12657), nullptr);
+                    player->RewardQuest(sObjectMgr->GetQuestTemplate(12657), false, player);
+                }
+                if (player->GetQuestStatus(12801) == QUEST_STATUS_NONE)//The Light of Dawn
+                {
+                    player->AddQuest(sObjectMgr->GetQuestTemplate(12801), nullptr);
+                    player->RewardQuest(sObjectMgr->GetQuestTemplate(12801), false, player);
+                }
+                if (player->GetTeam() == ALLIANCE && player->GetQuestStatus(13188) == QUEST_STATUS_NONE)//Where Kings Walk
+                    player->AddQuest(sObjectMgr->GetQuestTemplate(13188), nullptr);
+                else if (player->GetTeam() == HORDE && player->GetQuestStatus(13189) == QUEST_STATUS_NONE)//Saurfang's Blessing
+                    player->AddQuest(sObjectMgr->GetQuestTemplate(13189), nullptr);
+                if (player->GetTeam() == ALLIANCE)
+                    player->TeleportTo(0, -8833.37f, 628.62f, 94.00f, 1.06f);//Stormwind
+                else
+                    player->TeleportTo(1, 1569.59f, -4397.63f, 16.06f, 0.54f);//Orgrimmar
+                ObjectAccessor::SaveAllPlayers();//Save
+            }
+        }
+    }
+};
+
+void AddSC_skip_StarterArea()
+{
+    new Trinitycore_skip_deathknight_announce;
+    new Trinitycore_skip_deathknight;
+}
diff --git a/src/server/scripts/Custom/custom_script_loader.cpp b/src/server/scripts/Custom/custom_script_loader.cpp
index 9e5e9ba2bf..702789c280 100644
--- a/src/server/scripts/Custom/custom_script_loader.cpp
+++ b/src/server/scripts/Custom/custom_script_loader.cpp
@@ -19,6 +19,9 @@
 
 // The name of this function should match:
 // void Add${NameOfDirectory}Scripts()
+void AddSC_skip_StarterArea();
+
 void AddCustomScripts()
 {
+    AddSC_skip_StarterArea();
 }
diff --git a/src/server/worldserver/worldserver.conf.dist b/src/server/worldserver/worldserver.conf.dist
index 0a79664476..24728797c5 100644
--- a/src/server/worldserver/worldserver.conf.dist
+++ b/src/server/worldserver/worldserver.conf.dist
@@ -4089,3 +4089,43 @@ Metric.OverallStatusInterval = 1
 
 #
 ###################################################################################################
+
+###################################################################################################
+#Skip Death Knight Starting Area#
+#################################
+#
+#    Skip Death Knight Starting Area
+#        Enable this if you want to skip the Deathknight starting area
+#        Start with level 58 DK with Runeforging.
+#
+#    Skip.Deathknight.Starter.Announce.enable
+#         Description: Announces Module when player logs in.
+#         Default:    0 - (Disabled)
+#                     1 - (Enabled)
+
+Skip.Deathknight.Starter.Announce.enable = 1
+
+#    Skip.DeathKnight.Starter.Enable
+#         Active Module for all new accounts regardless of account level.
+#         Default:    0 - (Disabled)
+#                     1 - (Enabled)
+
+Skip.Deathknight.Starter.Enable = 1
+
+#    GM.Skip.Deathknight.Starter.Enable
+#         Active Module for all new accounts of a GM account level ONLY.
+#         Ensure you have Skip.Deathknight.Starter.Enable set to 0 if using this conf.
+#         Default:    0 - (Disabled)
+#                     1 - (Enabled)
+
+GM.Skip.Deathknight.Starter.Enable = 0
+
+#    Skip.Deathknight.Start.Level
+#         Adjusted Death Knight Start Level if module is used.
+#         Default:    58 - (Default)
+#
+
+Skip.Deathknight.Start.Level = 58
+
+#
+###################################################################################################
\ No newline at end of file